controllers/core/openstackversion_controller.go

```diff
@@ -0,0 +1,281 @@
+	}
+	instance.Status.Conditions.Init(&cl)
+
+	versionHelper, err := helper.NewHelper(
+		instance,
+		r.Client,
+		r.Kclient,
+		r.Scheme,
+		Log,
+	)
+	if err != nil {
+		Log.Error(err, "unable to create helper")
+		return ctrl.Result{}, err
+	}
+
+	// Always patch the instance status when exiting this function so we can persist any changes.
+	defer func() {
+		// update the Ready condition based on the sub conditions
+		if instance.Status.Conditions.AllSubConditionIsTrue() {
+			instance.Status.Conditions.MarkTrue(
+				condition.ReadyCondition, condition.ReadyMessage)
+		} else {
+			// something is not ready so reset the Ready condition
+			instance.Status.Conditions.MarkUnknown(
+				condition.ReadyCondition, condition.InitReason, condition.ReadyInitMessage)
+			// and recalculate it based on the state of the rest of the conditions
+			instance.Status.Conditions.Set(
+				instance.Status.Conditions.Mirror(condition.ReadyCondition))
+		}
+
+		err := versionHelper.PatchInstance(ctx, instance)
+		if err != nil {
+			_err = err
+			return
+		}
+	}()
+
+	instance.Status.Conditions.Set(condition.FalseCondition(
+		corev1beta1.OpenStackVersionInitialized,
+		condition.RequestedReason,
+		condition.SeverityInfo,
+		corev1beta1.OpenStackVersionInitializedReadyRunningMessage))
+
+	instance.Status.AvailableVersion = &envAvailableVersion
+	defaults := openstack.InitializeOpenStackVersionImageDefaults(ctx, envContainerImages)
+	// store the defaults for the currently available version
+	if instance.Status.ContainerImageVersionDefaults == nil {
+		instance.Status.ContainerImageVersionDefaults = make(map[string]*corev1beta1.ContainerDefaults)
+	}
+	instance.Status.ContainerImageVersionDefaults[envAvailableVersion] = defaults
+
+	// calculate the container images for the target version
+	if val, ok := instance.Status.ContainerImageVersionDefaults[instance.Spec.TargetVersion]; !ok {
+		Log.Info("Target version not found in defaults", "targetVersion", instance.Spec.TargetVersion)
+		return ctrl.Result{}, nil
+	} else {
+		instance.Status.ContainerImages = openstack.GetContainerImages(ctx, val, *instance)
+	}
+
+	instance.Status.Conditions.MarkTrue(
+		corev1beta1.OpenStackVersionInitialized,
+		corev1beta1.OpenStackVersionInitializedReadyMessage)
+
+	// greenfield deployment //FIXME check dataplane here too
+	if controlPlane == nil || controlPlane.Status.DeployedVersion == nil {
+		Log.Info("No controlplane or controlplane is not deployed")
+		return ctrl.Result{}, nil
+	}
+
+	// TODO minor update for OVN Dataplane in progress
+
+	// minor update for OVN Controlplane in progress
+	if instance.Status.DeployedVersion != nil && instance.Spec.TargetVersion != *instance.Status.DeployedVersion {
+		if !compareStringPointers(controlPlane.Status.ContainerImages.OvnControllerImage, instance.Status.ContainerImages.OvnControllerImage) ||
+			!controlPlane.Status.Conditions.IsTrue(corev1beta1.OpenStackVersionMinorUpdateOVNControlplane) {
```

Comment:
Where do we set OpenStackVersionMinorUpdateOVNControlplane on the controlPlane CR?

Reply:
I fixed this and just haven't pushed it yet as I was testing a new pattern. The actual condition here is OpenStackControlPlaneOVNReadyCondition